controllers/databasecluster_controller.go

```diff
@@ -1050,45 +1146,129 @@ func (r *DatabaseClusterReconciler) reconcilePXC(ctx context.Context, req ctrl.R
 	return r.Status().Update(ctx, database)
 }
 
-func (r *DatabaseClusterReconciler) createPGBackrestSecret(
+// Creates or updates the PG Backrest secret
+func CreatePGBackrestSecret(
 	ctx context.Context,
+	client client.Client,
+	scheme *runtime.Scheme,
 	database *everestv1alpha1.DatabaseCluster,
-	storageSecretName,
-	repoName,
+	pgbackrestS3Conf []byte,
 	pgBackrestSecretName string,
 ) (*corev1.Secret, error) {
-	s3StorageSecret := &corev1.Secret{}
-	err := r.Get(ctx, types.NamespacedName{Name: storageSecretName, Namespace: database.Namespace}, s3StorageSecret)
-	if err != nil {
-		return nil, err
-	}
-
-	pgbackrestS3Conf := fmt.Sprintf("[global]\n%s-s3-key=%s\n%s-s3-key-secret=%s\n",
-		repoName, s3StorageSecret.Data["AWS_ACCESS_KEY_ID"],
-		repoName, s3StorageSecret.Data["AWS_SECRET_ACCESS_KEY"],
-	)
 	pgBackrestSecret := &corev1.Secret{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      pgBackrestSecretName,
 			Namespace: database.Namespace,
 		},
 		Type: corev1.SecretTypeOpaque,
 		Data: map[string][]byte{
-			"s3.conf": []byte(pgbackrestS3Conf),
+			"s3.conf": pgbackrestS3Conf,
 		},
 	}
-	err = controllerutil.SetControllerReference(database, pgBackrestSecret, r.Scheme)
+	err := controllerutil.SetControllerReference(database, pgBackrestSecret, scheme)
 	if err != nil {
 		return nil, err
 	}
-	err = r.createOrUpdate(ctx, pgBackrestSecret, false)
+	err = CreateOrUpdate(ctx, client, pgBackrestSecret, false)
 	if err != nil {
 		return nil, err
 	}
 
 	return pgBackrestSecret, nil
 }
 
+// Returns the index of the backup storage in the pgbackrest repo list
+func GetBackupStorageIndexInPGBackrestRepo(
+	backupStorage *everestv1alpha1.BackupStorage,
+	repos []crunchyv1beta1.PGBackRestRepo,
+) int {
+	for idx, repo := range repos {
+		if repo.S3 == nil || repo.S3.Bucket != backupStorage.Spec.Bucket ||
+			repo.S3.Region != backupStorage.Spec.Region ||
+			repo.S3.Endpoint != backupStorage.Spec.EndpointURL {
+			continue
+		}
+		return idx
+	}
+
+	return -1
+}
+
+// Generates a PGBackrest repo for a given backup storage
+func GenPGBackrestRepo(
+	repoIdx int,
+	backupStorage *everestv1alpha1.BackupStorage,
+	backupSchedule *string,
+) (crunchyv1beta1.PGBackRestRepo, error) {
+	if repoIdx < 0 || repoIdx > 4 {
```

Comment:
Is there any reason the indexes are limited to 0-4?

Reply:
This is a limitation of the PG operator. I don't know if there's any reason for that limitation on their side or not.

Reply:
This just made me notice that we have a bug. The repo name must be repo[1-4] and we're allowing 0 as an input which isn't correct.

Reply:
I'm a bit worried about the 4 repos limit.
It can easily happen there will be more than 4 repos defined by the user in which case we automatically fail the reconcile loop and cannot make any more changes to the `DBCluster`.

Maybe we may need to handle this gracefully?
Or maybe create the repos only if they're really needed - e.g. when a backup is triggered?
Or maybe impose a limit in the backend so we prevent the users from having more than 4 repos?

What do you think?

Reply:
Yes, I completely agree with your points and have those concerns myself.
I actually wanted to have this as part of the extra validations we were planning on adding in the near future. I wanted to do this with admission webhooks like we explored in #34.
However Andrew shared this feedback about the webhooks so it's not entirely clear what path we will follow.
> Well. There are a couple of drawbacks to using webhooks:
>
> It makes the operator cluster wide which is a deal breaker for upstream operator customers and Peter Szczepaniak told that we will hit the same wall sooner or later 
> It requires an effort to improve CI/CD pipeline and it will make it even more unstable because webhooks are cluster-wide
> We need to enable deployments of an operator image into the EKS cluster for our pipeline which requires a lot of work to stabilize tests 
> We can face other unknown issues with our pipeline. 
> As an option, we can disable webhooks for pipeline runs yet we need to generate tls certificates anyway to make operator work

Reply:
So I would like to go with a 4th option which is pretty similar to your 3rd one but doing it in the operator instead by using admission webhooks.
i.e. impose a limit in the operator so we prevent the users from having more than 4 repos

If this isn't possible we'll have to go with the BE validation approach.

Reply:
Wouldn't this put us in the same situation as we're now - i.e. everest operator will fail to reconcile the db cluster and no changes to the cluster will be possible until the too many repos issue is resolved?

Reply:
No, the admission webhooks are run before storing the CR so if it fails the CR won't be changed. Think of it as a validation before committing a change to a DB, if the validation fails nothing gets changed.